renderer/markdown.go

```diff
@@ -0,0 +1,142 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+package renderer
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"strings"
+	"text/template"
+
+	"github.com/Masterminds/sprig"
+	"github.com/elastic/crd-ref-docs/config"
+	"github.com/elastic/crd-ref-docs/templates"
+	"github.com/elastic/crd-ref-docs/types"
+)
+
+type MarkdownRenderer struct {
+	conf *config.Config
+	*Functions
+}
+
+func NewMarkdownRenderer(conf *config.Config) (*MarkdownRenderer, error) {
+	baseFuncs, err := NewFunctions(conf)
+	if err != nil {
+		return nil, err
+	}
+	return &MarkdownRenderer{conf: conf, Functions: baseFuncs}, nil
+}
+
+func (m *MarkdownRenderer) Render(gvd []types.GroupVersionDetails) error {
+	funcMap := combinedFuncMap(funcMap{prefix: "markdown", funcs: m.ToFuncMap()}, funcMap{funcs: sprig.TxtFuncMap()})
+
+	var tpls fs.FS
+	if m.conf.TemplatesDir != "" {
+		tpls = os.DirFS(m.conf.TemplatesDir)
+	} else {
+		sub, err := fs.Sub(templates.Root, "markdown")
+		if err != nil {
+			return err
+		}
+		tpls = sub
+	}
+
+	tmpl, err := loadTemplate(tpls, funcMap)
+	if err != nil {
+		return err
+	}
+
+	f, err := createOutFile(m.conf.OutputPath, "out.md")
+	defer f.Close()
+
+	return tmpl.ExecuteTemplate(f, mainTemplate, gvd)
+}
+
+func (m *MarkdownRenderer) ToFuncMap() template.FuncMap {
+	return template.FuncMap{
+		"GroupVersionID":     m.GroupVersionID,
+		"RenderExternalLink": m.RenderExternalLink,
+		"RenderGVLink":       m.RenderGVLink,
+		"RenderLocalLink":    m.RenderLocalLink,
+		"RenderType":         m.RenderType,
+		"RenderTypeLink":     m.RenderTypeLink,
+		"SafeID":             m.SafeID,
+		"ShouldRenderType":   m.ShouldRenderType,
+		"TypeID":             m.TypeID,
+	}
+}
+
+func (m *MarkdownRenderer) ShouldRenderType(t *types.Type) bool {
+	return t != nil && (t.GVK != nil || len(t.References) > 0)
+}
+
+func (m *MarkdownRenderer) RenderType(t *types.Type) string {
+	var sb strings.Builder
+	switch t.Kind {
+	case types.MapKind:
+		sb.WriteString("object (")
+		sb.WriteString("keys:")
+		sb.WriteString(m.RenderTypeLink(t.KeyType))
+		sb.WriteString(", values:")
+		sb.WriteString(m.RenderTypeLink(t.ValueType))
+		sb.WriteString(")")
+	case types.ArrayKind, types.SliceKind:
+		sb.WriteString(m.RenderTypeLink(t.UnderlyingType))
+		sb.WriteString(" array")
+	default:
+		sb.WriteString(m.RenderTypeLink(t))
+	}
+
+	return sb.String()
+}
+
+func (m *MarkdownRenderer) RenderTypeLink(t *types.Type) string {
+	text := m.SimplifiedTypeName(t)
+
+	link, local := m.LinkForType(t)
+	if link == "" {
+		return text
+	}
+
+	if local {
+		return m.RenderLocalLink(text)
+	} else {
+		return m.RenderExternalLink(link, text)
+	}
+}
+
+func (m *MarkdownRenderer) RenderLocalLink(text string) string {
+	anchor := strings.ToLower(
+		strings.NewReplacer(
+			" ", "-",
+			".", "",
+			"/", "",
+			"(", "",
+			")", "",
+		).Replace(text),
+	)
+	return fmt.Sprintf("[%s](#%s)", text, anchor)
```

Comment:
What's the motivation for doing it differently from the asciidoc renderer which uses the `link` obtained from `LinkForType` in `RenderLocalLink`?

```diff
diff --git a/renderer/markdown.go b/renderer/markdown.go
index 04ceb50..7788f8c 100644
--- a/renderer/markdown.go
+++ b/renderer/markdown.go
@@ -114,23 +114,14 @@ func (m *MarkdownRenderer) RenderTypeLink(t *types.Type) string {
        }
 
        if local {
-               return m.RenderLocalLink(text)
+               return m.RenderLocalLink(link, text)
        } else {
                return m.RenderExternalLink(link, text)
        }
 }
 
-func (m *MarkdownRenderer) RenderLocalLink(text string) string {
-       anchor := strings.ToLower(
-               strings.NewReplacer(
-                       " ", "-",
-                       ".", "",
-                       "/", "",
-                       "(", "",
-                       ")", "",
-               ).Replace(text),
-       )
-       return fmt.Sprintf("[%s](#%s)", text, anchor)
+func (m *MarkdownRenderer) RenderLocalLink(link, text string) string {
+       return fmt.Sprintf("[%s](#%s)", text, link)
 }
 
 func (m *MarkdownRenderer) RenderExternalLink(link, text string) string {
@@ -138,5 +129,5 @@ func (m *MarkdownRenderer) RenderExternalLink(link, text string) string {
 }
 
 func (m *MarkdownRenderer) RenderGVLink(gv types.GroupVersionDetails) string {
-       return m.RenderLocalLink(gv.GroupVersionString())
+       return m.RenderLocalLink(m.GroupVersionID(gv), gv.GroupVersionString())
```

Reply:
@thbkrkr I'd really like to keep it that way. Any chance this can be merged as is?

Reply:
Apologies for the delay, yeah let's merge this as is.